File: vc-verifier/kotlin/vcverifier/src/main/java/io/mosip/vercred/vcverifier/utils/ValidationHelper.kt
package io.mosip.vercred.vcverifier.utils

import com.nimbusds.jose.JWSObject
import foundation.identity.jsonld.JsonLDObject
import info.weboftrust.ldsignatures.LdProof
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.ALGORITHM
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.ALGORITHMS_SUPPORTED
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.CREDENTIAL_SUBJECT
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.DESCRIPTION
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.ERROR_CODE_INVALID
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.ERROR_MESSAGE_ALGORITHM_NOT_SUPPORTED
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.ERROR_CODE_MISSING
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.ERROR_CREDENTIAL_SUBJECT_NON_NULL_OBJECT
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.ERROR_MESSAGE_DESCRIPTION
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.ERROR_INVALID_FIELD
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.ERROR_INVALID_URI
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.ERROR_MISSING_REQUIRED_FIELDS
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.ERROR_MESSAGE_NAME
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.ERROR_MESSAGE_PROOF_TYPE_NOT_SUPPORTED
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.ERROR_MESSAGE_TYPE_VERIFIABLE_CREDENTIAL
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.ID
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.ISSUER
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.JWS
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.LANGUAGE
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.NAME
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.PROOF
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.PROOF_TYPES_SUPPORTED
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.TYPE
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.VERIFIABLE_CREDENTIAL
import io.mosip.vercred.vcverifier.exception.ValidationException
import org.json.JSONArray
import org.json.JSONObject

class ValidationHelper {


    fun checkMandatoryFields(vcJsonObject: JSONObject, fields: List<String>) {

        for (field in fields) {
            val keys = field.split(".")
            var currentJson: JSONObject? = vcJsonObject

            for (key in keys) {
                if (currentJson != null && currentJson.has(key)) {
                    if (currentJson.get(key) is JSONObject) {
                        currentJson = currentJson.getJSONObject(key)
                    } else {
                        break
                    }
                } else {
                    val specificErrorCode = "$ERROR_CODE_MISSING${field.replace(".", "_").uppercase()}"
                    throw ValidationException( "$ERROR_MISSING_REQUIRED_FIELDS$field", specificErrorCode)
                }
            }
        }
    }

    fun validateProof(vcJsonString: String) {
        val vcJsonObject = JSONObject(vcJsonString)

        val vcJsonLdObject: JsonLDObject = JsonLDObject.fromJson(vcJsonString)
        val ldProof : LdProof = LdProof.getFromJsonLDObject(vcJsonLdObject)

        if(vcJsonObject.getJSONObject(PROOF).has(JWS)){
            val jwsToken: String = ldProof.jws
            val algorithmName: String = JWSObject.parse(jwsToken).header.algorithm.name
            if(jwsToken.isEmpty() || !ALGORITHMS_SUPPORTED.contains(algorithmName)){
                throw ValidationException( ERROR_MESSAGE_ALGORITHM_NOT_SUPPORTED , "${ERROR_CODE_INVALID}${ALGORITHM.uppercase()}")
            }
        }

        val ldProofType: String = ldProof.type
        if (!PROOF_TYPES_SUPPORTED.contains(ldProofType)) {
            throw ValidationException( ERROR_MESSAGE_PROOF_TYPE_NOT_SUPPORTED, "${ERROR_CODE_INVALID}${PROOF.uppercase()}_${TYPE.uppercase()}")
        }
    }

    fun validateId(vcJsonObject: JSONObject){
        if(vcJsonObject.has(ID)){
            if(!Util.isValidUri(vcJsonObject.getString(ID))){
                throw ValidationException("$ERROR_INVALID_URI$ID", "${ERROR_CODE_INVALID}${ID}")
            }
        }

    }

    fun validateType(vcJsonObject: JSONObject){

        if(vcJsonObject.has(TYPE)){
            vcJsonObject.optJSONArray(TYPE)?.let { types ->
                if (!Util.jsonArrayToList(types).contains(VERIFIABLE_CREDENTIAL)) {
                    throw ValidationException(ERROR_MESSAGE_TYPE_VERIFIABLE_CREDENTIAL, "${ERROR_CODE_INVALID}${TYPE.uppercase()}")
                }
            }
        }

    }

    fun validateIssuer(vcJsonObject: JSONObject){
        if(vcJsonObject.has(ISSUER)){
            val issuerId = Util.getId(vcJsonObject.get(ISSUER))
            if(issuerId == null || !Util.isValidUri(issuerId)) {
                throw ValidationException( "$ERROR_INVALID_URI${ISSUER}", "${ERROR_CODE_INVALID}${ISSUER.uppercase()}")
            }
        }
    }

    fun validateNameAndDescription(vcJsonObject: JSONObject) {

        val nameDescriptionList: List<Pair<String, String>> = listOf(
            NAME to ERROR_MESSAGE_NAME,
            DESCRIPTION to ERROR_MESSAGE_DESCRIPTION
        )

        nameDescriptionList.forEach { fieldPair ->
            if(vcJsonObject.has(fieldPair.first)){
                when (val fieldValue = vcJsonObject.get(fieldPair.first)) {
                    is String -> return
                    is JSONArray -> checkForLanguageObject(fieldValue, fieldPair)
                    else -> {
                        throw ValidationException(fieldPair.second, "${ERROR_CODE_INVALID}${fieldPair.first.uppercase()}")
                    }
                }
            }

        }

    }

    private fun checkForLanguageObject(
        nameOrDescriptionArray: JSONArray,
        fieldPair: Pair<String, String>
    ) {
        for (i in 0 until nameOrDescriptionArray.length()) {
            val nameObject = nameOrDescriptionArray.getJSONObject(i)
            if (!nameObject.has(LANGUAGE)) {
                throw ValidationException(fieldPair.second, "${ERROR_CODE_INVALID}${fieldPair.first.uppercase()}")
            }
        }
    }

    fun validateCredentialSubject(vcJsonObject: JSONObject) {
        val credentialSubject = vcJsonObject.get(CREDENTIAL_SUBJECT)
        validateJsonObjectOrArray(CREDENTIAL_SUBJECT, credentialSubject, ::validateSingleCredentialObject,
            ERROR_CREDENTIAL_SUBJECT_NON_NULL_OBJECT
        )
    }

    fun validateFieldsByIdAndType(vcJsonObject: JSONObject, fieldName: String, idMandatoryFields: List<String>) {
        val fieldValue = vcJsonObject.get(fieldName)
        validateJsonObjectOrArray(fieldName, fieldValue, { obj -> validateSingleObject(fieldName, obj, idMandatoryFields) }, "$ERROR_INVALID_FIELD$fieldName")
    }

    private fun validateJsonObjectOrArray(
        fieldName: String,
        value: Any,
        validator: (JSONObject) -> String,
        errorMessage: String
    ) {
        when (value) {
            is JSONArray -> {
                for (i in 0 until value.length()) {
                    val jsonObject = value.getJSONObject(i)
                    val result = validator(jsonObject)
                    if (result.isNotEmpty()) throw ValidationException(errorMessage, "${ERROR_CODE_INVALID}${fieldName.uppercase()}")
                }
            }
            is JSONObject -> validator(value)
            else -> throw ValidationException(errorMessage, "${ERROR_CODE_INVALID}${fieldName.uppercase()}")
        }
    }

    private fun validateSingleCredentialObject(credentialSubjectObject: JSONObject): String {
        if (credentialSubjectObject.has(ID) && !Util.isValidUri(credentialSubjectObject.optString(ID))) {
            throw ValidationException("$ERROR_INVALID_URI$CREDENTIAL_SUBJECT.$ID", "$ERROR_CODE_INVALID${CREDENTIAL_SUBJECT}${ID.uppercase()}")
        }
        return ""
    }

    private fun validateSingleObject(fieldName: String, fieldValueObject: JSONObject, idMandatoryFields: List<String>): String {
        if (!fieldValueObject.has(TYPE)) {
            throw ValidationException( "$ERROR_MISSING_REQUIRED_FIELDS$fieldName.$TYPE", "$ERROR_CODE_MISSING${fieldName.uppercase()}_${TYPE.uppercase()}")
        } else if (fieldValueObject.optString(TYPE).isNullOrBlank()) {
            throw ValidationException("$fieldName.$TYPE cannot be null or empty.", "$ERROR_CODE_INVALID${fieldName.uppercase()}_${TYPE.uppercase()}")
        }

        val isIDMandatoryField = idMandatoryFields.contains(fieldName)
        if (isIDMandatoryField) {
            if (!fieldValueObject.has(ID)) {
                throw ValidationException("$ERROR_MISSING_REQUIRED_FIELDS$fieldName.$ID", "$ERROR_CODE_MISSING${fieldName.uppercase()}_${ID.uppercase()}")
            } else if (fieldValueObject.optString(ID).isNullOrBlank()) {
                throw ValidationException("$fieldName.$ID cannot be null or empty.", "$ERROR_CODE_INVALID${fieldName.uppercase()}_${ID.uppercase()}")
            }
        }
        fieldValueObject.optString(ID).takeIf { it.isNotEmpty() }?.let { id ->
            if (!Util.isValidUri(id)) {
                throw ValidationException( "$ERROR_INVALID_URI$fieldName.$ID", "$ERROR_CODE_INVALID${fieldName.uppercase()}_${ID.uppercase()}")
            }
        }

        return ""
    }


}

File: vc-verifier/js/src/validator/ValidationHelper.js
import {ContextUrls, Errors, Fields, VERIFIABLE_CREDENTIAL} from "../constant/ValidatorConstants.js";
import {ValidationError} from "./ValidationError.js";
import jws from "jws";
import {getId, isNotNullOrEmpty, isObject, isValidURI} from "./Utils.js";

export const DataModel = {
    DATA_MODEL_1_1: "DATA_MODEL_1_1",
    DATA_MODEL_2_0: "DATA_MODEL_2_0"
};

export const getContextVersion = (credential) => {
    if (credential.hasOwnProperty(Fields.CONTEXT)) {
        const contextArray = credential[Fields.CONTEXT];
        const contextUrl = contextArray[0]; // Get the first element of the context array

        switch (contextUrl) {
            case ContextUrls.CREDENTIALS_CONTEXT_V1_URL:
                return DataModel.DATA_MODEL_1_1;
            case ContextUrls.CREDENTIALS_CONTEXT_V2_URL:
                return DataModel.DATA_MODEL_2_0;
            default:
                throw new ValidationError(Errors.ERROR_CONTEXT_FIRST_LINE)
        }
    } else {
        throw new ValidationError(`${Errors.ERROR_MISSING_REQUIRED_FIELDS}${Fields.CONTEXT}`)
    }
}

export const validateMandatoryFields = (credential, mandatoryFields) => {
    for (const field of mandatoryFields) {
        if (!(field in credential) || credential[field] === undefined) {

            throw new ValidationError(`${Errors.ERROR_MISSING_REQUIRED_FIELDS}${field}`);
        }
    }
}

export const validateCredentialSubject = (credential) => {
    const credentialSubject = credential[Fields.CREDENTIAL_SUBJECT];

    if (Array.isArray(credentialSubject)) {
        return credentialSubject.map(subject => _checkCredentialSubject({ subject }));
    } else if(isObject(credentialSubject)) {
        return _checkCredentialSubject({ subject: credentialSubject });
    } else {
        throw new ValidationError(`${Errors.ERROR_CREDENTIAL_SUBJECT_NON_NULL_OBJECT}`);
    }
}

export const _checkCredentialSubject = ({subject}) => {
    if(subject.id && !isValidURI(subject.id)) {
        throw new ValidationError(`${Errors.ERROR_INVALID_URI}${Fields.CREDENTIAL_SUBJECT}.${Fields.ID}`)
    }
}

export const validateIssuer = (credential) => {
    if(credential.hasOwnProperty(Fields.ISSUER)){
        const issuerId = getId(credential[Fields.ISSUER])
        if(issuerId == null || !isValidURI(issuerId)){
            throw new ValidationError(`${Errors.ERROR_INVALID_URI}${Fields.ISSUER}.${Fields.ID}`)
        }
    }
}

export const validateID = (credential) => {
    if(credential.hasOwnProperty(Fields.ID)){
        if(!isValidURI(credential[Fields.ID])){
            throw new ValidationError(`${Errors.ERROR_INVALID_URI}${Fields.ID}`)
        }
    }
}

export const validateType = (credential) => {
    if(credential.hasOwnProperty(Fields.TYPE)){
        if(!credential[Fields.TYPE] || !(credential[Fields.TYPE].length > 0) || (!credential[Fields.TYPE].includes(VERIFIABLE_CREDENTIAL))){
            throw new ValidationError(`${Errors.ERROR_TYPE_VERIFIABLE_CREDENTIAL}`)
        }
    }
}

export const validateProof = (credential) => {
    if(credential.hasOwnProperty(Fields.PROOF)){
        if(isNotNullOrEmpty(credential[Fields.PROOF])){
            const proofType = credential[Fields.PROOF][Fields.TYPE]
            if(credential[Fields.PROOF].hasOwnProperty(Fields.JWS)){
                if(isNotNullOrEmpty(proofType) && !PROOF_TYPES_SUPPORTED.includes(proofType)){
                    throw new ValidationError(`${Errors.ERROR_PROOF_TYPE_NOT_SUPPORTED}`)
                }
                const { header } = jws.decode(credential.proof.jws);
                if(!ALGORITHMS_SUPPORTED.includes(header.alg)){
                    throw new ValidationError(`${Errors.ERROR_ALGORITHM_NOT_SUPPORTED}`)
                }
            }
        }
    }
}

export const validateNameAndDescription = (credential)  => {
    const nameDescriptionList = [
        { field: `${Fields.NAME}`, error: `${Errors.ERROR_NAME}` },
        { field: `${Fields.DESCRIPTION}`, error:`${Errors.ERROR_DESCRIPTION}` },
    ];

    nameDescriptionList.forEach(({ field, error }) => {
        if (credential.hasOwnProperty(field)) {
            const fieldValue = credential[field];
            if (typeof fieldValue === 'string') {
                return;
            } else if (Array.isArray(fieldValue)) {
                checkForLanguageObject(fieldValue, error);
            } else {
                throw new ValidationError(error);
            }
        }
    });
}

export const checkForLanguageObject = (nameArray, errorMessage) => {
    nameArray.forEach((nameObject) => {
        if (!nameObject.hasOwnProperty(Fields.LANGUAGE)) {
            throw new ValidationError(errorMessage);
        }
    });
}


export const validateFieldsWithID = (credential, allFieldsWithTypeAndId, idMandatoryFields) => {
    allFieldsWithTypeAndId.forEach((fieldName) => {
        const fieldValue = credential[fieldName];
        if (fieldValue !== undefined) {
            if (Array.isArray(fieldValue)) {
                fieldValue.forEach((item) => validateSingleIDObject(fieldName, item, idMandatoryFields));
            } else if (typeof fieldValue === 'object' && fieldValue !== null) {
                validateSingleIDObject(fieldName, fieldValue, idMandatoryFields);
            } else {
                throw new ValidationError(`${Errors.ERROR_INVALID_FIELD}${fieldName}`);
            }
        }
    });
};

const validateSingleIDObject = (fieldName, fieldValueObject, idMandatoryFields) => {
    if (idMandatoryFields.includes(fieldName) && !fieldValueObject.hasOwnProperty(Fields.ID)) {
        throw new ValidationError(`${Errors.ERROR_MISSING_REQUIRED_FIELDS}${fieldName}.${Fields.ID}`);
    }

    const id = fieldValueObject[Fields.ID];
    if (id && !isValidURI(id)) {
        throw new ValidationError(`${Errors.ERROR_INVALID_URI}${fieldName}.${Fields.ID}`);
    }
};

export const validateFieldsWithType = (credential, typeMandatoryFields) => {
    typeMandatoryFields.forEach((fieldName) => {
        const fieldValue = credential[fieldName];
        if (fieldValue !== undefined) {
            if (Array.isArray(fieldValue)) {
                fieldValue.forEach((item) => validateSingleTypeObject(fieldName, item));
            } else if (typeof fieldValue === 'object' && fieldValue !== null) {
                validateSingleTypeObject(fieldName, fieldValue);
            } else {
                throw new ValidationError(`${Errors.ERROR_INVALID_FIELD}${fieldName}`);
            }
        }
    });
};

const validateSingleTypeObject = (fieldName, fieldValueObject) => {
    if (!fieldValueObject.hasOwnProperty(Fields.TYPE)) {
        throw new ValidationError(`${Errors.ERROR_MISSING_REQUIRED_FIELDS}${fieldName}.${Fields.TYPE}`);
    }
};



export const validationSuccess = (message) => ({
    verificationStatus: true,
    verificationErrorMessage: message ? message: ""
});

export const validationFailure = (error) => ({
    verificationStatus: false,
    verificationErrorMessage: error
});

const ALGORITHMS_SUPPORTED = ["PS256", "RS256"]
const PROOF_TYPES_SUPPORTED = ["RsaSignature2018"]

